code/modules/projectiles/gun.dm

```diff
@@ -461,10 +465,26 @@
 	return new_mode
 
 /obj/item/weapon/gun/attack_self(mob/user)
-	var/datum/firemode/new_mode = switch_firemodes(user)
-	if(new_mode)
-		playsound(src.loc, 'sound/weapons/guns/interact/selector.ogg', 100, 1)
-		user << SPAN_NOTICE("\The [src] is now set to [new_mode.name].")
+	if(zoom)
+		toggle_scope(zoom_factor)
+		return
+	var/list/options = list("firemode", "scope", "safety")
+	for(var/option in options)
+		options[option] = image(icon = 'icons/obj/gun_actions.dmi', icon_state = "[option]")
+	var/selected
+	selected = show_radial_menu(user, src, options, radius = 42)
+	if(!selected)
+		return
+	switch(selected)
+		if("firemode")
+			var/datum/firemode/new_mode = switch_firemodes(user)
+			if(new_mode)
+				playsound(src.loc, 'sound/weapons/guns/interact/selector.ogg', 100, 1)
+				user << SPAN_NOTICE("\The [src] is now set to [new_mode.name].")
```

Comment:
Use `to_chat` instead of `<<`. We plan to port goonchat one day, so adding even more code that uses `<<` is not something we want.